File: phases/14-graphics-and-visualization/17-visualization-d3-deck-gl-scientific-plotting/code/main.ts
// Visualization — D3, deck.gl, scientific plotting
// Phase 14 — Computer Graphics & Visualization
//
// D3-style data join and visual encoding pipeline in pure TypeScript.
// No browser, no D3 library — shows the core patterns that make D3 powerful.

// ── Data Types ──────────────────────────────────────────────────────────

type DataType = "categorical" | "ordinal" | "quantitative" | "temporal" | "spatial";

interface Datum {
  id: string;
  values: Record<string, number | string>;
}

interface VisualEncoding {
  channel: "x" | "y" | "color" | "size" | "shape" | "opacity";
  field: string;
  dataType: DataType;
  scale: Scale;
}

interface Scale {
  kind: "linear" | "log" | "ordinal" | "band";
  domain: [number, number] | string[];
  range: [number, number] | string[];
}

// ── Scale Functions ──────────────────────────────────────────────────────

function applyScale(scale: Scale, value: number | string): number | string {
  if (scale.kind === "linear" || scale.kind === "log") {
    const val = typeof value === "string" ? parseFloat(value) : value;
    const domain = scale.domain as [number, number];
    const range = scale.range as [number, number];
    if (scale.kind === "log") {
      const logVal = Math.log(val) / Math.log(domain[1]);
      const logMin = Math.log(domain[0]) / Math.log(domain[1]);
      const t = (logVal - logMin) / (1 - logMin);
      return range[0] + t * (range[1] - range[0]);
    }
    const t = (val - domain[0]) / (domain[1] - domain[0]);
    return range[0] + t * (range[1] - range[0]);
  }
  if (scale.kind === "ordinal") {
    const domain = scale.domain as string[];
    const range = scale.range as string[];
    const idx = domain.indexOf(String(value));
    return range[((idx % range.length) + range.length) % range.length];
  }
  if (scale.kind === "band") {
    const domain = scale.domain as string[];
    const range = scale.range as [number, number];
    const idx = domain.indexOf(String(value));
    const bandWidth = (range[1] - range[0]) / domain.length;
    return range[0] + idx * bandWidth + bandWidth * 0.15;
  }
  return 0;
}

// ── Perceptually Uniform Colormap (viridis-like) ──────────────────────

function viridisInterpolate(t: number): string {
  t = Math.max(0, Math.min(1, t));
  const r = Math.max(0, Math.min(255, Math.round((-1.26 * t + 0.32) * t * 255 + 0.27 * 255)));
  const g = Math.max(0, Math.min(255, Math.round((0.43 * t + 0.62) * 255)));
  const b = Math.max(0, Math.min(255, Math.round((1.0 * t + 0.14) * 255)));
  return `#${r.toString(16).padStart(2, "0")}${g.toString(16).padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
}

const TABLEAU_10: string[] = [
  "#4e79a7", "#f28e2b", "#e15759", "#76b7b2", "#59a14f",
  "#edc948", "#b07aa1", "#ff9da7", "#9c755f", "#bab0ac",
];

// ── Data Join Engine ─────────────────────────────────────────────────────

interface JoinResult<T> {
  enter: T[];
  update: T[];
  exit: T[];
}

/**
 * D3's data join pattern: given new data and existing items (both keyed),
 * partition into three disjoint sets:
 *
 *   ENTER — data present in new but missing from existing (add new elements)
 *   UPDATE — data present in both (modify existing elements)
 *   EXIT   — data present in existing but missing from new (remove elements)
 */
function dataJoin<T extends { id: string }>(
  newData: T[],
  existingData: T[]
): JoinResult<T> {
  const existingMap = new Map<string, T>();
  for (const item of existingData) {
    existingMap.set(item.id, item);
  }

  const newMap = new Map<string, T>();
  for (const item of newData) {
    newMap.set(item.id, item);
  }

  const enter: T[] = [];
  const update: T[] = [];
  const exit: T[] = [];

  for (const item of newData) {
    if (existingMap.has(item.id)) {
      update.push(item);
    } else {
      enter.push(item);
    }
  }

  for (const item of existingData) {
    if (!newMap.has(item.id)) {
      exit.push(item);
    }
  }

  return { enter, update, exit };
}

// ── Visual Encoding Pipeline ─────────────────────────────────────────────

interface RenderElement {
  id: string;
  type: "rect" | "circle" | "text" | "line" | "path";
  attributes: Record<string, number | string>;
  children?: RenderElement[];
}

/**
 * Transform raw data through the grammar of graphics pipeline:
 *
 *   Raw Data → Mapped Data → Scaled Data → Render Elements
 *
 * Each step is a pure function, composable and testable.
 */
function pipeDataToRender(
  data: Datum[],
  encodings: VisualEncoding[]
): RenderElement[] {
  return data.map((datum) => {
    const attrs: Record<string, number | string> = {};

    for (const enc of encodings) {
      const rawValue = datum.values[enc.field];
      const scaledValue = applyScale(enc.scale, rawValue);
      attrs[enc.channel] = scaledValue;
    }

    attrs.fill = attrs.color ?? TABLEAU_10[0];
    delete attrs.color;

    return {
      id: datum.id,
      type: encodings.find((e) => e.channel === "size") ? "circle" : "rect",
      attributes: attrs,
    };
  });
}

// ── SVG Renderer ──────────────────────────────────────────────────────────

function renderToSvg(elements: RenderElement[], width: number, height: number): string {
  const lines: string[] = [
    `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">`,
  ];

  for (const el of elements) {
    const attrStr = Object.entries(el.attributes)
      .map(([k, v]) => `${k}="${typeof v === "number" ? v.toFixed(1) : v}"`)
      .join(" ");
    if (el.type === "circle") {
      const cx = typeof el.attributes.x === "number" ? el.attributes.x : 0;
      const cy = typeof el.attributes.y === "number" ? el.attributes.y : 0;
      const r = typeof el.attributes.size === "number" ? el.attributes.size : 4;
      lines.push(`  <circle cx="${cx.toFixed(1)}" cy="${cy.toFixed(1)}" r="${r.toFixed(1)}" ${attrStr} />`);
    } else if (el.type === "rect") {
      const x = typeof el.attributes.x === "number" ? el.attributes.x : 0;
      const y = typeof el.attributes.y === "number" ? el.attributes.y : 0;
      const w = typeof el.attributes.width === "number" ? el.attributes.width : 20;
      const h = typeof el.attributes.height === "number" ? el.attributes.height : 20;
      lines.push(`  <rect x="${x.toFixed(1)}" y="${y.toFixed(1)}" width="${w.toFixed(1)}" height="${h.toFixed(1)}" ${attrStr} />`);
    } else if (el.type === "text") {
      lines.push(`  <text ${attrStr}>${el.id}</text>`);
    }
  }

  lines.push("</svg>");
  return lines.join("\n");
}

// ── Demonstration ────────────────────────────────────────────────────────

function demonstrateDataJoin(): void {
  console.log("=== Data Join Demo ===\n");

  const previousData: Datum[] = [
    { id: "A", values: { name: "Region A", revenue: 120 } },
    { id: "B", values: { name: "Region B", revenue: 80 } },
    { id: "C", values: { name: "Region C", revenue: 95 } },
  ];

  const newData: Datum[] = [
    { id: "A", values: { name: "Region A", revenue: 135 } },
    { id: "C", values: { name: "Region C", revenue: 100 } },
    { id: "D", values: { name: "Region D", revenue: 60 } },
    { id: "E", values: { name: "Region E", revenue: 45 } },
  ];

  const result = dataJoin(newData, previousData);

  console.log("Previous data:", previousData.map((d) => d.id).join(", "));
  console.log("New data:     ", newData.map((d) => d.id).join(", "));
  console.log("");
  console.log("ENTER  (add new elements):   ", result.enter.map((d) => `${d.id} (${d.values.name})`).join(", "));
  console.log("UPDATE (modify existing):     ", result.update.map((d) => `${d.id} (${d.values.name})`).join(", "));
  console.log("EXIT   (remove orphans):      ", result.exit.map((d) => `${d.id} (${d.values.name})`).join(", "));
  console.log("");
  console.log("This is the exact pattern D3's .data() + .enter()/.update/.exit() uses.");
  console.log("Only the minimal set of DOM operations is performed.\n");
}

function demonstrateEncodingPipeline(): void {
  console.log("=== Visual Encoding Pipeline Demo ===\n");

  const data: Datum[] = [
    { id: "North", values: { region: "North", revenue: 120, growth: 0.15 } },
    { id: "South", values: { region: "South", revenue: 80, growth: 0.08 } },
    { id: "East", values: { region: "East", revenue: 95, growth: 0.22 } },
    { id: "West", values: { region: "West", revenue: 60, growth: 0.31 } },
  ];

  const encodings: VisualEncoding[] = [
    {
      channel: "x",
      field: "region",
      dataType: "categorical",
      scale: { kind: "band", domain: ["North", "South", "East", "West"], range: [50, 590] },
    },
    {
      channel: "y",
      field: "revenue",
      dataType: "quantitative",
      scale: { kind: "linear", domain: [0, 140], range: [380, 20] },
    },
    {
      channel: "color",
      field: "region",
      dataType: "categorical",
      scale: { kind: "ordinal", domain: ["North", "South", "East", "West"], range: TABLEAU_10 },
    },
    {
      channel: "size",
      field: "growth",
      dataType: "quantitative",
      scale: { kind: "linear", domain: [0, 0.4], range: [3, 12] },
    },
  ];

  const elements = pipeDataToRender(data, encodings);

  console.log("Data → Encoded Elements:");
  for (const el of elements) {
    console.log(`  ${el.id}: x=${(el.attributes.x as number)?.toFixed?.(1) ?? "?"}, ` +
      `y=${(el.attributes.y as number)?.toFixed?.(1) ?? "?"}, ` +
      `fill=${el.attributes.fill}, size=${typeof el.attributes.size === "number" ? el.attributes.size.toFixed(1) : "?"}`);
  }
  console.log("");

  const svg = renderToSvg(elements, 640, 400);
  console.log("Generated SVG (" + svg.length + " bytes):");
  console.log(svg.substring(0, 300) + "...\n");
}

function demonstrateClevelandHierarchy(): void {
  console.log("=== Cleveland's Encoding Hierarchy ===\n");

  const hierarchy: [string, string, number][] = [
    ["Position (common scale)", "Most accurate — encode your most important variable here", 5],
    ["Position (identical scales)", "Very accurate — e.g., scatter plot dots", 5],
    ["Length", "Accurate — bar chart bars, but shorter bars look thinner", 4],
    ["Angle / Slope", "Moderate — pie charts are hard to read for this reason", 3],
    ["Area", "Poor — circles/squares encode area, but perception is non-linear", 2],
    ["Volume", "Very poor — 3D charts distort comparisons", 1],
    ["Color saturation", "Weak — useful for binning but not precision", 1],
    ["Color hue", "Weakest for quantity — great for categories, terrible for numbers", 1],
  ];

  for (const [encoding, implication, stars] of hierarchy) {
    const bar = "★".repeat(stars) + "☆".repeat(5 - stars);
    console.log(`  ${bar}  ${encoding}`);
    console.log(`         ${implication}`);
    console.log("");
  }
}

function demonstrateColorSafety(): void {
  console.log("=== Perceptually Uniform Colormaps ===\n");

  console.log("Viridis at 0%, 25%, 50%, 75%, 100%:");
  for (const t of [0, 0.25, 0.5, 0.75, 1.0]) {
    console.log(`  ${t.toFixed(2)} → ${viridisInterpolate(t)}`);
  }
  console.log("");

  console.log("Why NOT rainbow/jet:");
  console.log("  - Yellow-green band appears as a sharp boundary (luminance spike)");
  console.log("  - Blue-to-cyan gradual change looks like nothing (low luminance contrast)");
  console.log("  - Creates artificial structure where none exists in the data");
  console.log("  - Fails 3 types of colorblindness simulation");
  console.log("");

  console.log("Colorblind-safe rules:");
  console.log("  1. Never use red+green alone (use red+blue or orange+blue)");
  console.log("  2. Vary lightness, not just hue");
  console.log("  3. For quantitative: use viridis, magma, inferno, or plasma");
  console.log("  4. For categorical: use Tableau 10 or ColorBrewer qualitative");
}

// ── Entry Point ──────────────────────────────────────────────────────────

function main(): void {
  console.log("Visualization — D3, deck.gl, Scientific Plotting");
  console.log("Phase 14 — Computer Graphics & Visualization\n");

  demonstrateDataJoin();
  demonstrateEncodingPipeline();
  demonstrateClevelandHierarchy();
  demonstrateColorSafety();

  console.log("\n=== Chart Type Selection Guide ===\n");
  const guide: [string, string][] = [
    ["Categorical comparison", "Bar chart (position on common scale)"],
    ["Distribution of 1 variable", "Histogram (position + length)"],
    ["Relationship of 2 variables", "Scatter plot (position × position)"],
    ["Temporal trend", "Line chart (position over time)"],
    ["Geospatial density", "Heatmap (position + color)"],
    ["3+ variable composition", "Small multiples (facet by category)"],
    ["> 100K points on a map", "deck.gl scatter-plot layer (GPU)"],
    ["Rich interaction + < 10K points", "D3 (DOM-based, full control)"],
  ];
  for (const [scenario, chart] of guide) {
    console.log(`  ${scenario.padEnd(35)} → ${chart}`);
  }
}

main();

File: phases/14-graphics-and-visualization/17-visualization-d3-deck-gl-scientific-plotting/code/main.py
"""
Visualization — D3, deck.gl, scientific plotting
Phase 14 — Computer Graphics & Visualization

Generate SVG visualizations from raw data using only Python stdlib.
No matplotlib. No external dependencies. Pure string construction.

Outputs bar chart, scatter plot, and heatmap as SVG files.
"""

import json
import math
import os
from dataclasses import dataclass, field
from typing import Callable


# ── Data Types and Structures ────────────────────────────────────────────

@dataclass
class DataPoint:
    x: float
    y: float
    label: str = ""
    category: str = ""
    value: float = 0.0


@dataclass
class HeatmapCell:
    row: int
    col: int
    value: float


@dataclass
class BarDatum:
    label: str
    value: float


@dataclass
class Scale:
    domain_min: float
    domain_max: float
    range_min: float
    range_max: float

    def __call__(self, value: float) -> float:
        if self.domain_max == self.domain_min:
            return (self.range_min + self.range_max) / 2
        t = (value - self.domain_min) / (self.domain_max - self.domain_min)
        return self.range_min + t * (self.range_max - self.range_min)


def svg_begin(width: int, height: int) -> str:
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" '
        f'width="{width}" height="{height}" '
        f'viewBox="0 0 {width} {height}">\n'
    )


def svg_end() -> str:
    return "</svg>\n"


def svg_rect(x: float, y: float, w: float, h: float,
             fill: str = "#4e79a7", stroke: str = "none",
             stroke_width: float = 0, rx: float = 0) -> str:
    parts = [f'  <rect x="{x:.1f}" y="{y:.1f}" width="{w:.1f}" height="{h:.1f}"']
    parts.append(f' fill="{fill}"')
    if stroke != "none":
        parts.append(f' stroke="{stroke}" stroke-width="{stroke_width:.1f}"')
    if rx > 0:
        parts.append(f' rx="{rx:.1f}"')
    parts.append("/>")
    return "".join(parts)


def svg_circle(cx: float, cy: float, r: float,
               fill: str = "#4e79a7", opacity: float = 1.0) -> str:
    return (
        f'  <circle cx="{cx:.1f}" cy="{cy:.1f}" r="{r:.1f}" '
        f'fill="{fill}" opacity="{opacity:.2f}"/>'
    )


def svg_text(x: float, y: float, text: str,
             size: int = 12, anchor: str = "middle",
             fill: str = "#333") -> str:
    escaped = text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    return (
        f'  <text x="{x:.1f}" y="{y:.1f}" font-size="{size}" '
        f'text-anchor="{anchor}" fill="{fill}">{escaped}</text>'
    )


def svg_line(x1: float, y1: float, x2: float, y2: float,
             stroke: str = "#ccc", width: float = 1.0) -> str:
    return (
        f'  <line x1="{x1:.1f}" y1="{y1:.1f}" '
        f'x2="{x2:.1f}" y2="{y2:.1f}" '
        f'stroke="{stroke}" stroke-width="{width:.1f}"/>'
    )


def svg_group(attrs: str = "") -> str:
    return f"  <g {attrs}>" if attrs else "  <g>"


def svg_group_end() -> str:
    return "  </g>"


# ── Perceptually Uniform Colormap (viridis-like) ────────────────────

def viridis_interpolate(t: float) -> str:
    t = max(0.0, min(1.0, t))
    r = int((-1.26 * t + 0.32) * t * 255 + 0.27 * 255) & 0xFF
    g = int((0.43 * t + 0.62) * 255) & 0xFF
    b = int((1.0 * t + 0.14) * 255) & 0xFF
    r = max(0, min(255, r))
    g = max(0, min(255, g))
    b = max(0, min(255, b))
    return f"#{r:02x}{g:02x}{b:02x}"


def magma_interpolate(t: float) -> str:
    t = max(0.0, min(1.0, t))
    r = int(min(255, (0.001 + 1.69 * t) * 255)) & 0xFF
    g = int(min(255, (0.0 + 0.63 * t * t) * 255)) & 0xFF
    b = int(min(255, (0.09 + 0.67 * t) * 255)) & 0xFF
    r = max(0, min(255, r))
    g = max(0, min(255, g))
    b = max(0, min(255, b))
    return f"#{r:02x}{g:02x}{b:02x}"


# ── Bar Chart ────────────────────────────────────────────────────────────

TABLEAU_10 = [
    "#4e79a7", "#f28e2b", "#e15759", "#76b7b2", "#59a14f",
    "#edc948", "#b07aa1", "#ff9da7", "#9c755f", "#bab0ac",
]


def generate_bar_chart_svg(data: list[BarDatum],
                           width: int = 640,
                           height: int = 400,
                           title: str = "Bar Chart",
                           palette: list[str] | None = None) -> str:
    colors = palette or TABLEAU_10
    margin_l, margin_r, margin_t, margin_b = 60, 20, 40, 50
    plot_w = width - margin_l - margin_r
    plot_h = height - margin_t - margin_b

    if not data:
        return svg_begin(width, height) + svg_text(width // 2, height // 2, "No data") + svg_end()

    max_val = max(d.value for d in data)
    y_scale = Scale(0, max_val * 1.15, plot_h, 0)
    y_scale_inv = Scale(0, max_val * 1.15, 0, plot_h)

    n = len(data)
    band_w = plot_w / n
    bar_w = band_w * 0.7

    parts = [svg_begin(width, height)]

    for i in range(5):
        grid_val = max_val * 1.15 * (i + 1) / 5
        gy = margin_t + y_scale(grid_val)
        parts.append(svg_line(margin_l, gy, margin_l + plot_w, gy, "#ddd", 0.5))
        parts.append(svg_text(margin_l - 8, gy + 4, f"{grid_val:.0f}", 10, "end"))

    parts.append(svg_line(margin_l, margin_t + plot_h, margin_l + plot_w, margin_t + plot_h, "#333", 1.5))
    parts.append(svg_line(margin_l, margin_t, margin_l, margin_t + plot_h, "#333", 1.5))

    for i, datum in enumerate(data):
        bx = margin_l + i * band_w + (band_w - bar_w) / 2
        bar_h = y_scale_inv(datum.value)
        by = margin_t + plot_h - bar_h
        color = colors[i % len(colors)]
        parts.append(svg_rect(bx, by, bar_w, bar_h, fill=color, rx=2))
        parts.append(svg_text(
            margin_l + i * band_w + band_w / 2,
            margin_t + plot_h + 18,
            datum.label, 11
        ))
        parts.append(svg_text(bx + bar_w / 2, by - 6, f"{datum.value:.0f}", 9))

    parts.append(svg_text(width // 2, 20, title, 14, "middle", "#222"))

    desc = f"Bar chart titled '{title}' with {n} categories"
    parts.insert(1, f'  <desc>{desc}</desc>')

    parts.append(svg_end())
    return "\n".join(parts)


# ── Scatter Plot ─────────────────────────────────────────────────────────

def generate_scatter_svg(data: list[DataPoint],
                         width: int = 640,
                         height: int = 480,
                         title: str = "Scatter Plot") -> str:
    margin_l, margin_r, margin_t, margin_b = 60, 20, 40, 50
    plot_w = width - margin_l - margin_r
    plot_h = height - margin_t - margin_b

    if not data:
        return svg_begin(width, height) + svg_text(width // 2, height // 2, "No data") + svg_end()

    xs = [d.x for d in data]
    ys = [d.y for d in data]
    x_scale = Scale(min(xs), max(xs), 0, plot_w)
    y_scale = Scale(min(ys), max(ys), plot_h, 0)

    categories = sorted(set(d.category for d in data if d.category)) or ["default"]
    cat_colors = {c: TABLEAU_10[i % len(TABLEAU_10)] for i, c in enumerate(categories)}

    parts = [svg_begin(width, height)]
    parts.append(f'  <desc>Scatter plot titled "{title}" with {len(data)} points</desc>')

    for i in range(5):
        gy = margin_t + plot_h * i / 4
        parts.append(svg_line(margin_l, gy, margin_l + plot_w, gy, "#eee", 0.5))
        y_val = min(ys) + (max(ys) - min(ys)) * (1 - i / 4)
        parts.append(svg_text(margin_l - 8, gy + 4, f"{y_val:.1f}", 10, "end"))

    for i in range(5):
        gx = margin_l + plot_w * i / 4
        parts.append(svg_line(gx, margin_t, gx, margin_t + plot_h, "#eee", 0.5))
        x_val = min(xs) + (max(xs) - min(xs)) * i / 4
        parts.append(svg_text(gx, margin_t + plot_h + 18, f"{x_val:.1f}", 10))

    parts.append(svg_line(margin_l, margin_t + plot_h, margin_l + plot_w, margin_t + plot_h, "#333", 1.5))
    parts.append(svg_line(margin_l, margin_t, margin_l, margin_t + plot_h, "#333", 1.5))

    for pt in data:
        cx = margin_l + x_scale(pt.x)
        cy = margin_t + y_scale(pt.y)
        cat = pt.category or "default"
        color = cat_colors.get(cat, TABLEAU_10[0])
        parts.append(svg_circle(cx, cy, 4, fill=color, opacity=0.75))

    legend_y = margin_t + 10
    for cat, color in cat_colors.items():
        parts.append(svg_circle(margin_l + plot_w + 10, legend_y, 4, fill=color))
        parts.append(svg_text(margin_l + plot_w + 20, legend_y + 4, cat, 10, "start"))
        legend_y += 18

    parts.append(svg_text(width // 2, 20, title, 14, "middle", "#222"))
    parts.append(svg_end())
    return "\n".join(parts)


# ── Heatmap ──────────────────────────────────────────────────────────────

def generate_heatmap_svg(cells: list[HeatmapCell],
                         rows: int, cols: int,
                         row_labels: list[str] | None = None,
                         col_labels: list[str] | None = None,
                         width: int = 560,
                         height: int = 480,
                         title: str = "Heatmap") -> str:
    margin_l, margin_r, margin_t, margin_b = 80, 40, 40, 40
    plot_w = width - margin_l - margin_r
    plot_h = height - margin_t - margin_b

    if not cells:
        return svg_begin(width, height) + svg_text(width // 2, height // 2, "No data") + svg_end()

    cell_w = plot_w / cols
    cell_h = plot_h / rows
    values = [c.value for c in cells]
    v_min, v_max = min(values), max(values)
    v_scale = Scale(v_min, v_max, 0, 1)

    parts = [svg_begin(width, height)]
    parts.append(f'  <desc>Heatmap titled "{title}" with {rows}x{cols} grid</desc>')

    cell_map: dict[tuple[int, int], HeatmapCell] = {(c.row, c.col): c for c in cells}

    for r in range(rows):
        for c in range(cols):
            cell = cell_map.get((r, c))
            if cell is None:
                continue
            t = v_scale(cell.value)
            color = viridis_interpolate(t)
            x = margin_l + c * cell_w
            y = margin_t + r * cell_h
            parts.append(svg_rect(x, y, cell_w, cell_h, fill=color, stroke="#fff", stroke_width=1))
            if cell_w > 30 and cell_h > 18:
                text_color = "#fff" if t > 0.5 else "#333"
                parts.append(svg_text(
                    x + cell_w / 2, y + cell_h / 2 + 4,
                    f"{cell.value:.1f}", 9, "middle", text_color
                ))

    if row_labels:
        for r, label in enumerate(row_labels[:rows]):
            y = margin_t + r * cell_h + cell_h / 2
            parts.append(svg_text(margin_l - 8, y, label, 10, "end"))

    if col_labels:
        for c, label in enumerate(col_labels[:cols]):
            x = margin_l + c * cell_w + cell_w / 2
            parts.append(svg_text(x, margin_t + plot_h + 18, label, 9))

    bar_x = margin_l + plot_w + 8
    bar_w = 16
    for i in range(20):
        t = i / 19
        color = viridis_interpolate(t)
        by = margin_t + plot_h - t * plot_h
        bh = plot_h / 19 + 1
        parts.append(svg_rect(bar_x, by, bar_w, bh, fill=color))
    parts.append(svg_text(bar_x + bar_w + 4, margin_t + 6, f"{v_max:.1f}", 9, "start"))
    parts.append(svg_text(bar_x + bar_w + 4, margin_t + plot_h + 4, f"{v_min:.1f}", 9, "start"))

    parts.append(svg_text(width // 2, 20, title, 14, "middle", "#222"))
    parts.append(svg_end())
    return "\n".join(parts)


# ── Sample Data Generators ───────────────────────────────────────────────

def sample_bar_data() -> list[BarDatum]:
    return [
        BarDatum("North", 120),
        BarDatum("South", 80),
        BarDatum("East", 95),
        BarDatum("West", 60),
        BarDatum("Central", 135),
    ]


def sample_scatter_data() -> list[DataPoint]:
    import random
    rng = random.Random(42)
    data: list[DataPoint] = []
    for cat, cx, cy, spread in [("Cluster A", 3, 5, 1.2), ("Cluster B", 8, 9, 1.0), ("Cluster C", 5, 12, 0.8)]:
        for _ in range(25):
            x = rng.gauss(cx, spread)
            y = rng.gauss(cy, spread * 0.9)
            data.append(DataPoint(x=x, y=y, category=cat))
    return data


def sample_heatmap_data(rows: int = 8, cols: int = 10) -> tuple[list[HeatmapCell], int, int]:
    cells: list[HeatmapCell] = []
    for r in range(rows):
        for c in range(cols):
            value = 10 * math.sin(r * 0.5) * math.cos(c * 0.6) + 15 + r * 0.5
            cells.append(HeatmapCell(row=r, col=c, value=round(value, 1)))
    return cells, rows, cols


# ── Main ─────────────────────────────────────────────────────────────────

def main() -> None:
    out_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "outputs")
    os.makedirs(out_dir, exist_ok=True)

    bar_svg = generate_bar_chart_svg(sample_bar_data(), title="Regional Sales (units)")
    with open(os.path.join(out_dir, "bar_chart.svg"), "w") as f:
        f.write(bar_svg)
    print(f"Wrote bar_chart.svg ({len(bar_svg)} bytes)")

    scatter_svg = generate_scatter_svg(sample_scatter_data(), title="Cluster Distribution")
    with open(os.path.join(out_dir, "scatter_plot.svg"), "w") as f:
        f.write(scatter_svg)
    print(f"Wrote scatter_plot.svg ({len(scatter_svg)} bytes)")

    cells, rows, cols = sample_heatmap_data()
    row_labels = [f"Row {i}" for i in range(rows)]
    col_labels = [f"C{i}" for i in range(cols)]
    heatmap_svg = generate_heatmap_svg(
        cells, rows, cols,
        row_labels=row_labels,
        col_labels=col_labels,
        title="Signal Intensity Heatmap",
    )
    with open(os.path.join(out_dir, "heatmap.svg"), "w") as f:
        f.write(heatmap_svg)
    print(f"Wrote heatmap.svg ({len(heatmap_svg)} bytes)")


if __name__ == "__main__":
    main()